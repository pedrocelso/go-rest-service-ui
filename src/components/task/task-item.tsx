import React, { Component } from 'react';
import moment from 'moment-timezone'
import { Task } from '../../services/task';

interface TaskItemProps {
  task: Task;
}

export const formatTime = (time: number): string => {
  const date = new Date(time)
  return moment(date, moment.tz.guess()).format("YYYY/MM/DD, h:mm:ss a")
}

export class TaskItem extends Component<TaskItemProps> {
  render() {
    const {task} = this.props
    const timezone = moment.tz.guess()

    return (
      <div className="list-group-item d-flex row">
        <div className="col-md-7">
          <b>{task.name} - {task.description}</b>
        </div>
        <div className="col-md-2">
          {formatTime(task.creationTime)}
        </div>
        <div className="col-md-2">
          {formatTime(task.updateTime)}
        </div>
        <div className="col-md-1">
          <span className="badge badge-primary badge-pill">0</span>
        </div>
      </div>
    )
  }
}
