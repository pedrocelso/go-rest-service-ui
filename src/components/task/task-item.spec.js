import React from 'react';
import {shallow} from 'enzyme'
import sinon from 'sinon'

import {formatTime, TaskItem} from './task-item'

const baseTask = {
  id: 123,
  name: `POI quest service`,
  description: `Do A POI quest service`,
  creationTime: 1556924666404,
  updateTime: 1556924566404,
  incidentsCount: 1,
  pendingIncidentsCount: 1
}

const deauthenticate = sinon.fake()

describe(`<TaskItem />`, () => {
  it(`Should render the task without errors`, () => {
    const service = sinon.fake()
    const component = shallow(
      <TaskItem task={baseTask} service={service} deauthenticate={deauthenticate} />,
    );

    expect(component).toMatchSnapshot();
  });

  it(`Should render the task with no incidents without errors`, () => {
    const service = sinon.fake()
    const task = {
      ...baseTask,
      incidentsCount: 0,
      pendingIncidentsCount: 0
    }
    const component = shallow(
      <TaskItem task={task} service={service} deauthenticate={deauthenticate} />,
    );

    expect(component).toMatchSnapshot();
  });

  it(`.formatTime() - Should format date properly`, () => {
    const formattedDate = formatTime(1556924666404)
    expect(formattedDate).toEqual(`2019/05/03, 4:04:26 pm`)
  });

  it(`.formatTime() - Should retutn n/a for 0 millis`, () => {
    const formattedDate = formatTime(0)
    expect(formattedDate).toEqual(`n/a`)
  });

  it(`Should load incidents from service`, () => {
    const service = {
      getIncidents: (_) => ({
        fork: (rej, res) => {
          res([
            { id: 10, name: `test`, description: `test@test.com` },
            { id: 11, name: `test1`, description: `test1@test.com` },
          ])
        }
      })
    }
    const component = shallow(
      <TaskItem task={baseTask} service={service} deauthenticate={deauthenticate} />,
    );

    component.setState({loadingIncidents: true})

    component.instance().loadIncidents()();
    const state = component.instance().state

    expect(state.loadingIncidents).toBeFalsy()
    expect(state.incidents.length).toEqual(2)
    expect(state.incidents[0].id).toEqual(10)
    expect(state.incidents[0].name).toEqual(`test`)
    expect(state.incidents[0].description).toEqual(`test@test.com`)
    expect(state.incidents[1].id).toEqual(11)
    expect(state.incidents[1].name).toEqual(`test1`)
    expect(state.incidents[1].description).toEqual(`test1@test.com`)
  });

  it(`Should deauthenticate with 401 response`, () => {
    const service = {
      getIncidents: (_) => ({
        fork: (rej, res) => {
          rej({statusCode: 401})
        }
      })
    }

    const deauthenticate = sinon.spy()

    const component = shallow(
      <TaskItem task={baseTask} service={service} deauthenticate={deauthenticate} />,
    );

    component.setState({loadingIncidents: true})
    component.instance().loadIncidents()();
    expect(deauthenticate.calledOnce).toBeTruthy()
  });

  it(`Should change loading state when status code is not 401`, () => {
    const service = {
      getIncidents: (_) => ({
        fork: (rej, res) => {
          rej({statusCode: 503})
        }
      })
    }

    const deauthenticate = sinon.spy()

    const component = shallow(
      <TaskItem task={baseTask} service={service} deauthenticate={deauthenticate} />,
    );

    component.setState({loadingIncidents: true})
    component.instance().loadIncidents()();
    expect(deauthenticate.called).toBeFalsy()
    expect(component.instance().state.loadingIncidents).toBeFalsy()
  });

  it(`Should not load anythong if task has no incidents`, () => {
    const forkSpy = sinon.spy()
    const service = {
      getIncidents: (_) => ({
        fork: forkSpy
      })
    }

    const deauthenticate = sinon.spy()
    const task = {
      ...baseTask,
      incidentsCount: 0
    }

    const component = shallow(
      <TaskItem task={task} service={service} deauthenticate={deauthenticate} />,
    );

    component.instance().loadIncidents()();
    expect(forkSpy.called).toBeFalsy()
    expect(component.instance().state.loadingIncidents).toBeFalsy()
  });
});