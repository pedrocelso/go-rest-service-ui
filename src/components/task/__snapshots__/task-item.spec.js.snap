// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskItem /> Should render the task without errors 1`] = `
<TaskItem
  task={
    Object {
      "creationTime": 1556924666404,
      "description": "Do A POI quest service",
      "id": 123,
      "name": "POI quest service",
      "updateTime": 1556924566404,
    }
  }
>
  <div
    className="list-group-item d-flex row"
  >
    <div
      className="col-md-7"
    >
      <b>
        POI quest service
         - 
        Do A POI quest service
      </b>
    </div>
    <div
      className="col-md-2"
    >
      2019/05/03, 4:04:26 pm
    </div>
    <div
      className="col-md-2"
    >
      2019/05/03, 4:02:46 pm
    </div>
    <div
      className="col-md-1"
    >
      <span
        className="badge badge-primary badge-pill"
      >
        0
      </span>
    </div>
  </div>
</TaskItem>
`;
